Add .row-breakout to enable layouts that spill into the page margins

lyubomir-popov opened the pull request:
Add a .row-breakout class that allows content to bleed into the page margins (up to a certain point).

QA:
 -Go to /docs/patterns/grid#breaking-out-of-the-grid
- Proof read copy, suggest improvements if something doesn't make sense
- Try importing the plugin with modified arguments, as described in the copy, ensure it works
- QA the following examples on different screen sizes: 
1) left aside: 
docs/examples/patterns/grid/grid-breakout-left-aside

2) right aside:
/docs/examples/patterns/grid/grid-breakout-right-aside

- Try to break it in different ways and report back :)

webteam-app commented:
Demo starting at https://vanilla-framework-3504.demos.haus

lyubomir-popov commented:
Slight misalignment happening at the point font size switches from 16 to 18, something to fix:

![image](https://user-images.githubusercontent.com/2741678/105230045-ba8f3b80-5b5c-11eb-9e8d-5d718c2737fc.png)

lyubomir-popov commented:
@bartaz another way to avoid a wrapper: introduce `row-*` classes same as we have `col-*` classes. This feels the most native way to use css grid:

![image](https://user-images.githubusercontent.com/2741678/105497666-11108d00-5cb7-11eb-915d-91419acf52c4.png)

bartaz commented:
> @bartaz another way to solve the stretching of items in each column without resorting to a wrapper and rigid predefined structures: introduce `row-*` classes same as we have `col-*` classes.

Ok. Yes, that could work. But this means you need to define how many rows you want from the start in HTML, it can be quite limiting. If content changes it may need adjusting.

lyubomir-popov commented:
> But this means you need to define how many rows you want from the start in HTML, it can be quite limiting. If content changes it may need adjusting.

to an extent, yes - but you don't need to be exact - giving it more rows wouldn't hurt, e.g. you could just give it 999 rows

lyubomir-popov commented:
- Added the cve table from https://ubuntu.com/security/cve as an example:
![image](https://user-images.githubusercontent.com/2741678/105745295-09125080-5f36-11eb-9c5e-1cc03c6de607.png)
- Added a grid where the left edge aligns to the fixed width grid, but the content in the aside and the content area can span as much space as it needs. Currently set an overall limit of 120rem:
![image](https://user-images.githubusercontent.com/2741678/105745469-3b23b280-5f36-11eb-864b-8687995f2a5d.png)

lyubomir-popov commented:
@bartaz could you take a look at this - I think all the functionality I want is in there, now it's a question of where to move stuff and what to do with the d3 chart

bartaz commented:
I'm not sure if we should extend the 'default' grid example with this. It was meant to be simple default 12 grid demo, that is also embedded in CodePen on grid docs page.
Throwing all of the 16 wide grid examples there makes it much harder to understand in docs. And included d3 doesn't work in CodePen at all (if we want it embedded we will need to find another way of including it in example).

I would suggest keeping default grid example as it was (just 12 column demo) and create separate example for 16 col grid. Ideally, one basic (just red boxes) that can be embedded, and separate one with exploration of different use cases and specific examples, that will be maybe part of Layout or Templates examples group and not embedded in the docs.

bartaz commented:
There seem to be some bugs in the example, especially on smaller screen size - the overlapping part is cut:

<img width="1141" alt="Screenshot 2021-01-27 at 11 00 33" src="https://user-images.githubusercontent.com/83575/105975696-89989480-608f-11eb-8f2a-3fb5dc39cfea.png">

Also, there is some HTML comment leaked at the end of the page:
<img width="1132" alt="Screenshot 2021-01-27 at 11 00 43" src="https://user-images.githubusercontent.com/83575/105975769-a339dc00-608f-11eb-86f5-cc838cbe7c79.png">

And 'S' on the left side of the table:

<img width="1125" alt="Screenshot 2021-01-27 at 11 00 49" src="https://user-images.githubusercontent.com/83575/105975798-ac2aad80-608f-11eb-8f40-7015f3c64f61.png">

bartaz commented on templates/static/js/chart.js:
```diff
@@ -0,0 +1,100 @@
+function responsivefy(svg) {
```
Unfortunately it seems that adding these files in the same folder as example HTML is a bit of a problem.

There are two options here

- we keep them in `/static/js` (maybe just make it `static/js/examples` to know better what they are used for.
- or keep them with example HTML, but include them in HTML with `{% include %}` rather than `<script src="" />`, but this means that it will be inlined in HTML. it may be fine for short scripts, but we likely don't want whole d3 library inlined in the example

So, I guess keeping it in `static/js/examples` would be better for now, unless we have better ideas? @sowasred2012

bartaz commented on scss/_patterns_grid.scss:
```diff
@@ -179,4 +192,31 @@
+  // Ways to break out of the 12-col grid
+
+  // Aside / main aligned to the left edge of the fixed-width grid
+  .row-new {
```
Why is it called `-new`? Does it replace `.row`? Is it a variant?
It may be "new" now, but will not be in couple months.

lyubomir-popov replied:
I actualy renamed it to row-breakout since  Ilast pushed - happy to change it to something else if you prefer

bartaz replied:
Anything more semantic than 'new' would be better.

On the other hand, as I mentioned in my other comment rant (sorry for making it that long 😉 ) I'm starting to think that maybe trying to reuse the grid naming here is not that well, because it seems to be behaving different from the grid.

bartaz commented:
TBH, from the examples, I don't know how this new grid is supposed to work.

I look at this:

<img width="1593" alt="Screenshot 2021-01-27 at 16 22 01" src="https://user-images.githubusercontent.com/83575/106012666-e90c9980-60bb-11eb-803f-0c36efbab265.png">

and I don't know, why is 'filters' sidebar on top different than from the one on bottom. Why are the other boxes called 'items' not columns? Why the bottom ones are so small as 1 col? (the top ones get so small as well on other screen width) - do we ever want to have anything that small? Can I control width of that? When I add `col-2` to the item it seems to be doing something (making it "2 columns" wide), but these item columns are not the same as standard grid columns.

I know that it's a quite a complicated piece of layout. But it does get me thinking how should we document it, so it can be used in practice. Designers need to know how to use it - what can be built with that, what can't be built with that. And then developers need to know how to use it - and there is a load of new class names here, new concepts of wrapper elements needed around certain parts of the grid, etc. At this point I'm not sure if we should try to reuse 'grid' naming here, because (even that it's kind of extension of the grid) it is build  and behaves so different from standard grid, that it seems that it's more confusing than helpful.

lyubomir-popov commented:
@bartaz after talking to @spencerbygraves we decided to drop the 16-col grid - as the narrowness of the aside makes it useless. So the bottom variant is gone. I've also renamed he areas to aside (which can go left/right) and main, much simpler. Let me iron out a few things and I'll push again.

lyubomir-popov commented:
@bartaz moved the combined examples to a template: https://vanilla-framework-3504.demos.haus/docs/examples/templates/grid-breakout-variants

bartaz commented:
@lyubomir-popov On a large screens the 'items' in new grid seem to have reasonable size (to be a card in store for example), but once you go down the size they all become 1 col with making them very narrow to whatever content they may have.

![grid-breakout](https://user-images.githubusercontent.com/83575/106114572-fd9a7180-614f-11eb-9da1-0bf879849382.gif)

Is this expected? Do we intend to ever use such small items in practice, or should it always be used at least as col-2? It doesn't seem clear from the example. And because large screen items seem to have acceptable size, the smaller screen going down to col-1 width looks broken.

lyubomir-popov commented:
That's a bug, will fi

On Thu, Jan 28, 2021 at 9:04 AM Bartek Szopka <notifications@github.com>
wrote:

> @lyubomir-popov <https://github.com/lyubomir-popov> On a large screens
> the 'items' in new grid seem to have reasonable size (to be a card in store
> for example), but once you go down the size they all become 1 col with
> making them very narrow to whatever content they may have.
>
> [image: grid-breakout]
> <https://user-images.githubusercontent.com/83575/106114572-fd9a7180-614f-11eb-9da1-0bf879849382.gif>
>
> Is this expected? Do we intend to ever use such small items in practice,
> or should it always be used at least as col-2? It doesn't seem clear from
> the example. And because large screen items seem to have acceptable size,
> the smaller screen going down to col-1 width looks broken.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical-web-and-design/vanilla-framework/pull/3504#issuecomment-768907485>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAU5LLRVSPVI3CTEJMS4QOLS4ESDFANCNFSM4WLGBI2Q>
> .
>

bartaz commented:
Does the height of 'main content' with items need to be constrained by the size of the sidebar?

What if sidebar is short and there is lots of items?
<img width="1751" alt="Screenshot 2021-01-28 at 10 06 21" src="https://user-images.githubusercontent.com/83575/106115188-c9738080-6150-11eb-8a7e-e762a914f371.png">

What if sidebar is very long, and there is not that many items (filtered search results)?
<img width="1781" alt="Screenshot 2021-01-28 at 10 07 38" src="https://user-images.githubusercontent.com/83575/106115271-e1e39b00-6150-11eb-8b55-eefbddceda88.png">

bartaz commented:
After this has been distilled down to simply being a list of items with a sidebar it seems a bit confusing to make it part of the grid and call it `row`.

In standard grid we have row, and its children are columns. In this new breakout layout the wrapper (that we call 'row-breakout') can actually be multiple rows of items. There are no real columns (on markup level you don't control the width of the items, they automatically take all available space and have some min-width). It feels to me that trying to call it similar to regular grid may make it more confusing than helpful. We shouldn't make developers or designers fell that it behaves similar to 12 col grid.

I don't know how to call this, but it probably should be a layout king of a thing. `l-breakout`? `l-grid-breakout`?

lyubomir-popov commented:
> I don't know how to call this, but it probably should be a layout king of a thing. `l-breakout`? `l-grid-breakout`?

That's what I initially had in mind, and I agree. @sowasred2012 @spencerbygraves any ideas for what to call this new layout? It may not always have cards, or a sidebar - for example it mihgt just have a cve table. Does breakout sound fitting? Or maybe something like `fluid`?

bartaz commented:
> It may not always have cards, or a sidebar - for example it mihgt just have a cve table.

Currently all the examples show sidebar and multiple cards.
If we have other usecases in mind we should show them in examples as well. Maybe keep the existing "sidebar left" "sidebar right" examples for embedding in docs, but have a separate "variants" example that shows different possibilities (like cards + sidebar, large table, etc).

bartaz commented on scss/_patterns_grid.scss:
```diff
@@ -98,23 +75,23 @@
       @extend %span-full-grid-on-mobile;
       @extend %span-full-grid-on-desktop;
-      @extend %display-block;
+      // @extend %display-block;
```
Is this intended? Other columns do extend display block. Why is this one commented?

lyubomir-popov replied:
not intended, thanks - fixed

bartaz commented on scss/_patterns_grid.scss:
```diff
@@ -27,44 +27,21 @@
   @media (max-width: $threshold-4-6-col) {
     @for $i from $grid-columns-small through 1 {
-      .#{$grid-small-col-prefix}#{$i} {
+      .row .#{$grid-small-col-prefix}#{$i} {
```
I guess the reason to add `.row` here is to make sure `col-X` class names work only in standard row context but not on new breakout layout items?

lyubomir-popov replied:
yes, otherwise it will have unpredictable effect. Just a driveby bug fix.

lyubomir-popov commented on scss/_settings_grid.scss:
```diff
@@ -30,6 +29,7 @@ $grid-margin-widths: (
-$threshold-12-col-plus: $grid-max-width + $aside-autofit-width;
+$l-fluid-breakout-max-width: $grid-max-width !default;
+$l-fluid-breakout-aside-width: 13rem !default;
+$l-fluid-breakout-main-width: 13rem !default;
```
just a misleading var name, this controls the min width of children of main. Would `$l-fluid-breakout-main-child-width` make it clearer?

lyubomir-popov commented on scss/_patterns_grid.scss:
```diff
@@ -158,18 +158,26 @@
       @supports (display: grid) {
         display: grid;
-        grid-template-columns: minmax($aside-autofit-width, 1fr) minmax($main-items-autofit-width, $breakout-autofit-width) minmax($aside-autofit-width, 1fr);
+        grid-template-columns: minmax($l-fluid-breakout-aside-width, 1fr) minmax($l-fluid-breakout-main-width, $l-fluid-breakout-max-width) minmax(
```
I'll add some sort of fallback

lyubomir-popov replied:
added fallback

lyubomir-popov commented:
@bartaz can you review the fallback stuff?

lyubomir-popov commented:
<img width="1680" alt="Screenshot 2021-02-08 at 15 06 16" src="https://user-images.githubusercontent.com/2741678/107237978-3d4e3c80-6a1f-11eb-83d4-1638967a44a7.png">
<img width="1391" alt="Screenshot 2021-02-08 at 15 05 49" src="https://user-images.githubusercontent.com/2741678/107237995-417a5a00-6a1f-11eb-8341-9fafdc62f90d.png">
<img width="683" alt="Screenshot 2021-02-08 at 14 57 06" src="https://user-images.githubusercontent.com/2741678/107238000-4212f080-6a1f-11eb-9d6c-b1c57cc39bbc.png">
<img width="945" alt="Screenshot 2021-02-08 at 14 56 35" src="https://user-images.githubusercontent.com/2741678/107238003-42ab8700-6a1f-11eb-9286-24957628c4a9.png">
<img width="1281" alt="Screenshot 2021-02-08 at 14 56 26" src="https://user-images.githubusercontent.com/2741678/107238007-43441d80-6a1f-11eb-8aed-6c9fa75d635f.png">

bartaz commented:
@lyubomir-popov I think we should restructure the documentation for it a bit.

Because we technically put it as part of the layouts, instead of adding very long section to already long Grid page, I think it should be a separate "Fluid grid breakout" page in "Layouts" section of the docs.
In similar way we could move examples for it out of the patterns folder to layouts folder and change their titles from "Grid" to "Fluid breakout".

bartaz commented:
@lyubomir-popov would it be possible to clean up commits a bit? Squash them into a couple bigger commits or just into one. Right now there is almost 40 commits in this PR, many of them have messages like "wip" "typos" etc. If at any time in future we try to figure out where the changes come from it will be a hell to navigate through those.

bartaz commented on scss/_patterns_grid.scss:
```diff
@@ -180,3 +157,204 @@
   }
 }
+
+@mixin vf-l-fluid-breakout(
```
As this is a separate mixin and a layout outside of the grid it should be in a separate file called `_layouts_something`.

lyubomir-popov replied:
done

bartaz commented on scss/_patterns_grid.scss:
```diff
@@ -180,3 +157,204 @@
+    }
+  }
+
+  %l-fluid-breakout__full-and-main-ie11-fallback {
```
As it seems that we want to allow including `vf-l-fluid-breakout` multiple times in same SCSS with different params, defining %placeholders inside of it may cause issues. Especially that these placeholders seem to use the parameters of the mixin (`$l-fluid-breakout-aside-width`, `$l-fluid-breakout-main-child-width`), because these placeholders will all be added to the same global scope under the same name.

Not sure if that's supported in SCSS, but worth trying, to do the same with placeholder name as with class name `.l-fluid-breakout#{$suffix}` so placeholder would be `%l-fluid-breakout__full-and-main-ie11-fallback#{suffix}` (but I have no idea if that's possible).

lyubomir-popov replied:
seems to work fine, nothing breaks

bartaz commented on templates/docs/examples/patterns/card/_header-and-footer.html:
```diff
@@ -0,0 +1,11 @@
+<div class="p-card">
+  <!-- <div class="p-strip is-shallow u-no-padding--top">
```
Should be removed if not needed.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/card/_header-and-footer.html:
```diff
@@ -0,0 +1,11 @@
+<div class="p-card">
+  <!-- <div class="p-strip is-shallow u-no-padding--top">
+  </div> -->
+    <img class="p-card__thumbnail" src="https://res.cloudinary.com/canonical/image/fetch/f_auto,q_auto,fl_sanitize,c_fill,w_64,h_64/https://api.charmhub.io/api/v1/media/download/charm_B1WVq07SAhAIBJ72odJQ9fNBCFXk7QYT_icon__ce0ecd4c4fb29e9a8ff84971fbbba8552a822d53717632f2aa5ecbce9dfea108.png" alt="landscape-server" class="p-card__thumbnail" loading="lazy" width="32" height="32"/>
```
Link to some image on asset server, not an external service.

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-no-aside.html:
```diff
@@ -0,0 +1,20 @@
+{% block content %}
+
+<style>
+  .viz-grid-item, .viz-aside {
```
This seems to be repeated a lot in these examples. Maybe extract it to separate file and `{% include %}` here?

lyubomir-popov replied:
done

bartaz commented on templates/docs/patterns/grid.md:
```diff
@@ -80,3 +105,101 @@ For more information see [Customising Vanilla](/docs/customising-vanilla/) in yo
 
 For more information [view the grid design spec](https://github.com/ubuntudesign/vanilla-design/tree/master/Grid), which includes the specification in markdown format and a PNG image.
+
+## Breaking out of the grid
```
As mentioned separately, would be good to split it into separate page.

bartaz replied:
But keep a section on Grid page that links there.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/templates/_card-fixed-width.html:
```diff
@@ -0,0 +1,33 @@
+{% block standalone_css %}patterns_card{% endblock %}
+
+{% block test_content %}
+<!-- <div class="p-card">
```
Remove stuff that is not needed.

lyubomir-popov replied:
removed

bartaz replied:
@lyubomir-popov Doesn't seem removed

lyubomir-popov replied:
oh sorry wrong file, this is actually not used, removed the entire file

bartaz commented on templates/docs/examples/patterns/side-navigation/_grid.html:
```diff
@@ -0,0 +1,83 @@
+        <a class="p-side-navigation__link">Concepts & terms</a>
+      </li>
+    </ul>
+    <!-- <ul class="p-side-navigation__list">
```
Remove if not needed.

lyubomir-popov replied:
removed entire file

bartaz commented on templates/docs/examples/patterns/grid/_cve-table.html:
```diff
@@ -0,0 +1,539 @@
+      <th>Priority</th>
+      <th>Package</th>
+      <th>
+        <div class="icon-container__text"> 14.04 ESM </div>
```
I guess that these class names `icon-container__text` `cve-table-cell-priority`, etc are just copy paste from the CVE page, but served as part of Vanilla example they seem a bit confusing as they don't provide any styling so there is no use for them in the example.

lyubomir-popov replied:
I've cleaned up the entire file

bartaz commented on scss/standalone/patterns_grid-breakout.scss:
```diff
@@ -0,0 +1,10 @@
+
+@import '../patterns_grid';
+@include vf-p-grid;
+@include vf-l-fluid-breakout;
```
You need to `@import '../layouts_fluid-breakout';` because it's moved to separate file now.

lyubomir-popov replied:
done

bartaz commented on scss/standalone/patterns_grid-breakout-full.scss:
```diff
@@ -0,0 +1,6 @@
+
+@import '../patterns_grid';
+@include vf-p-grid;
+@include vf-l-fluid-breakout(13rem, 100%, 0, '--single-main');
```
You need to `@import '../layouts_fluid-breakout';` because it's moved to separate file now.

lyubomir-popov replied:
done

clagom commented:
Apart from the comment on the inverted examples showing the same layout, I think this is working very well @lyubomir-popov. I would +1 this UX wise now, and then ask UX to have a look to iterate on it where necessary and to implement UX considerations and guidelines around your work. Great work as usual :)

lyubomir-popov commented:
Thanks a lot @clagom!

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-right-aside.html:
```diff
@@ -0,0 +1,27 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - aside on the right{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / Wide - aside on the right"?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-no-aside.html:
```diff
@@ -0,0 +1,19 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - no aside{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / ..."?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-left-aside.html:
```diff
@@ -0,0 +1,66 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - aside on the left{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / ..."?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-full.html:
```diff
@@ -0,0 +1,17 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - main spanning 3 columns{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / ..."?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-full--cve-table.html:
```diff
@@ -0,0 +1,34 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - CVE table{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / ..."?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/grid-breakout-cards-and-aside.html:
```diff
@@ -0,0 +1,37 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Grid / Wide - cards and aside{% endblock %}
```
Please rename it, as it's not Grid example anymore ("Fluid breakout / ..."?).

Also, because it's not a pattern move the file from `/templates/docs/examples/patterns/grid` to `/templates/docs/examples/layouts/...`.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/_cve-table.html:
```diff
@@ -0,0 +1,323 @@
+<table>
```
When moving examples from `patterns` to `layouts` this should be moved as well.

lyubomir-popov replied:
done

bartaz commented on templates/docs/examples/patterns/grid/_aside.html:
```diff
@@ -0,0 +1,64 @@
+<form>
```
When moving examples from `patterns` to `layouts` this should be moved as well.

lyubomir-popov replied:
done

bartaz reviewed (changes requested):
Please also move examples from "Patterns" to "Layouts"

bartaz commented on templates/docs/patterns/grid.md:
```diff
@@ -66,6 +91,10 @@ Please note, specifying a value that exceeds the available number of columns wil
 
+### Breaking out of the grid
+
+In some cases, there might be a good reason to break out of the constraints of a 12 column grid and allow content to bleed into the page margins. Vanilla provides a separate [layout](/docs/layouts/fluid-breakout) for this purpose
```
Missing full stop, and more descriptive link.

```suggestion
In some cases, there might be a good reason to break out of the constraints of a 12 column grid and allow content to bleed into the page margins. Vanilla provides a separate [fluid breakout layout](/docs/layouts/fluid-breakout) for this purpose.
```

bartaz commented on templates/docs/layouts/fluid-breakout.md:
```diff
@@ -0,0 +1,103 @@
+
+The `.l-fluid-breakout` serves as a wrapper, and allows for an aside and a main area. On screens smaller than `$breakpoint--large`, it acts as a single column. On larger screens it switches to a 3 column layout - a central area and two aside areas on either side.
+
+The wrapper aims to align as much as possible with the 12 column grid. On smaller screens, than is limited to ensuring the padding of both layouts match. Once the window width becomes large enough, the left edges of the regular 12 column grid and the central column of the fluid breakout layout also align. For this to be possible, two conditions need to be met:
```
Typo?

```suggestion
The wrapper aims to align as much as possible with the 12 column grid. On smaller screens, that is limited to ensuring the padding of both layouts match. Once the window width becomes large enough, the left edges of the regular 12 column grid and the central column of the fluid breakout layout also align. For this to be possible, two conditions need to be met:
```

bartaz commented on templates/docs/layouts/fluid-breakout.md:
```diff
@@ -0,0 +1,103 @@
+
+The wrapper aims to align as much as possible with the 12 column grid. On smaller screens, than is limited to ensuring the padding of both layouts match. Once the window width becomes large enough, the left edges of the regular 12 column grid and the central column of the fluid breakout layout also align. For this to be possible, two conditions need to be met:
+
+- <p>the central area's width needs to match the 12 column grid's width (we achieve this by setting the central area's max-width to the same value as the 12 column grid's max-width)</p>
```
Why is this using paragraphs inside the list?

lyubomir-popov replied:
not needed in narrow layouts, I think I added it when I experimented with the wider area - as list items have no max width. Removing as not currently needed.

bartaz reviewed (commented):
Couple questions/suggestions on the docs.

bartaz commented on scss/_layouts_fluid-breakout.scss:
```diff
@@ -0,0 +1,199 @@
+    justify-content: flex-start;
+
+    @media (min-width: $l-fluid-breakout-3-col-threshold) {
+      // display: flex;
```
Remove if not needed
```suggestion
```

bartaz commented on scss/_layouts_fluid-breakout.scss:
```diff
@@ -0,0 +1,199 @@
+      @extend %l-fluid-breakout__main-and-full-common;
+
+      @media (min-width: $breakpoint-large) {
+        @supports (display: grid) {
```
Would it make sense to wrap everything (that is not IE fallback) into `@supports (display: grid)`?
Right not it seems you separately use `@supports` in every single selector/query.

lyubomir-popov replied:
some of the styles - like padding - shouldn't be wrapped in supports, that's why it is wrapped in smaller pieces

bartaz commented on scss/standalone/patterns_grid-breakout.scss:
```diff
@@ -0,0 +1,13 @@
+@import '../base';
```
Is this file still needed or was it renamed to `layouts_fluid-breakout`?

lyubomir-popov replied:
removed

bartaz commented on scss/standalone/patterns_grid-breakout-full.scss:
```diff
@@ -0,0 +1,12 @@
+@import '../base';
```
Is this file still needed or was it renamed to `layouts_fluid-breakout-full`?

lyubomir-popov replied:
no removed

bartaz reviewed (commented):
Couple small comments/suggestions on the code.

bartaz commented on templates/docs/layouts/fluid-breakout.md:
```diff
@@ -0,0 +1,103 @@
+
+A couple of examples of where this layout might be useful. Both examples include a section with the regular 12 column grid for comparison.
+
+A large table that would require truncation if fitted within the regular 12 column grid:
```
Because this example relies on custom parameters passed to fluid layout it would be good to mention that in the docs, so it is clear that this example required different params than default built in layout.

lyubomir-popov replied:
added

lyubomir-popov replied:
reworded

bartaz commented on templates/docs/examples/patterns/card/_header-and-footer.html:
```diff
@@ -0,0 +1,9 @@
+<div class="p-card">
+  <img class="p-card__thumbnail" src="https://res.cloudinary.com/canonical/image/fetch/f_auto,q_auto,fl_sanitize,c_fill,w_64,h_64/https://api.charmhub.io/api/v1/media/download/charm_B1WVq07SAhAIBJ72odJQ9fNBCFXk7QYT_icon__ce0ecd4c4fb29e9a8ff84971fbbba8552a822d53717632f2aa5ecbce9dfea108.png" alt="landscape-server" class="p-card__thumbnail" loading="lazy" width="32" height="32" />
```
I feel like I commented on this before - please change the image to something from asset server, not claudinary.

lyubomir-popov replied:
fixed

bartaz commented on scss/_layouts_fluid-breakout.scss:
```diff
@@ -0,0 +1,216 @@
+      }
+    }
+
+    .l-fluid-breakout__full {
```
Do we need to have a separate class name for `__full`? "Full" sounds like a modifier, not an element. It seems that it's just a variant of `__main`. So maybe we could have it as `l-fluid-breakout__main.is-full-width`?

I think with layouts we should aim for a common structure of HTML and modify the layout by modifier classes, not by changing class names on children.

So in this case I think our default structure should be:

```html
<div class="l-fluid-breakout">
   <div class="l-fluid-breakout__aside"></div>
   <div class="l-fluid-breakout__main">
       <div class="l-fluid-breakout__item">
   </div>
</div>
```

Where `__aside` is optional (and can be before or after).
And `__main` is always there, but can be modified. If there is no aside, you add `.no-aside` to it.
If you want it to take full width you add `is-full-width` (instead of replacing it with `__full`).

Also, for an "item" class name we should use one consistent `l-fluid-breakout__item`, and not distinguish between `main-item` and `full-item`. From HTML point of view they are the same, we should be able to change their style based on parent modifier.

Would that be possible?

bartaz commented on scss/_layouts_fluid-breakout.scss:
```diff
@@ -0,0 +1,207 @@
+      }
+    }
+
+    .l-fluid-breakout__main-item {
```
For simplicity, can we rename it to `l-fluid-breakout__item`? I don't think we need to nest under main .

```suggestion
    .l-fluid-breakout__item {
```

bartaz reviewed (approved):
Thanks a lot!
